iree/turbine/kernel/boo/ops/graph.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2025 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from hashlib import sha1
+from typing import Any, Callable, Tuple, Sequence
+
+import torch
+from torch.fx.graph_module import GraphModule
+
+from torch._functorch.aot_autograd import aot_export_joint_simple
+from torch._functorch.partitioners import default_partition
+from torch.autograd import Function
+
+from iree.compiler.extras.fx_importer import FxImporter
+
+from .library import *
+from .utils import is_boo_backward_enabled, get_arg_spec_name
+from ..runtime import get_launchable
+from ....runtime.launch import Launchable
+from ....dynamo.passes import turbine_cpu_pass_pipeline
+from ....transforms.general.custom_op_expansion import ExpandCustomOpsPass
+from ....support.logging import aot_logger as logger
+
+__all__ = [
+    "get_io_from_gm",
+    "get_schema",
+    "get_custom_graph_op",
+    "define_custom_graph_op",
+    "get_mlir_module",
+    "get_autograd_function",
+]
+
+
+def get_io_from_gm(gm):
+    """Returns input nodes and output TensorMetadata from the graph module."""
+
+    inputs = []
+    meta_outputs = []
+    for node in gm.graph.nodes:
+        if node.op == "placeholder":
+            inputs.append(node.target)
+        if node.op == "output":
+            meta_outputs.extend(
+                [
+                    val.meta.get("tensor_meta", None) if val is not None else None
+                    for val in node.args[0]
+                ]
+            )
+    return inputs, meta_outputs
+
+
+def get_schema(inputs, outputs):
+    """Generate a schema from the result of `get_io_from_gm`."""
+
+    ret_ty = "Tensor?" if any([o is None for o in outputs]) else "Tensor"
+    schema = "("
+    schema += ", ".join([f"Tensor {inp}" for inp in inputs])
+    schema += ") -> ("
+    schema += ", ".join([ret_ty for _ in outputs])
+    schema += ")"
+    return schema
+
+
+def get_mlir_module(gm: GraphModule) -> Any:
+    """Generates torch-mlir IR from a graph module."""
+    sample_args = tuple(
+        [n.meta.get("val") for n in gm.graph.nodes if n.op == "placeholder"]
```

Comment:
I see this snippet repeatedly. It's short, but may be worth factoring out into a named function for readability.